src/gt4py/next/type_system_2/types.py

```diff
@@ -0,0 +1,257 @@
+            required_width = max(self.width, other.width)
+            return FloatType(required_width)
+        return None
+
+
+@dataclasses.dataclass
+class TupleType(Type, traits.FromTrait, traits.FromImplicitTrait):
+    elements: list[Type]
+
+    def __init__(self, elements: list[Type]):
+        self.elements = elements
+
+    def __str__(self):
+        elements = ", ".join(str(element) for element in self.elements)
+        return f"tuple[{elements}]"
+
+    def is_constructible_from(self, ty: Type) -> bool:
+        if not isinstance(ty, TupleType):
+            return False
+        if len(self.elements) != len(ty.elements):
+            return False
+        return all(traits.is_convertible(oth_el, el) for el, oth_el in zip(self.elements, ty.elements))
+
+    def is_implicitly_constructible_from(self, ty: Type) -> bool:
+        if not isinstance(ty, TupleType):
+            return False
+        if len(self.elements) != len(ty.elements):
+            return False
+        return all(traits.is_implicitly_convertible(oth_el, el) for el, oth_el in zip(self.elements, ty.elements))
+
+
+@dataclasses.dataclass
+class StructType(Type):
+    fields: list[tuple[str, Type]]
+
+    def __init__(self, fields: list[tuple[str, Type]]):
+        self.fields = fields
+
+    def __str__(self):
+        elements = ", ".join(f"{field[0]}: {str(field[1])}" for field in self.fields)
+        return f"struct[{elements}]"
+
+
+@dataclasses.dataclass
+class FunctionParameter:
+    ty: Type
+    name: str
+    positional: bool
+    keyword: bool
+
+
+@dataclasses.dataclass
+class FunctionType(Type, traits.CallableTrait):
+    parameters: list[FunctionParameter]
+    result: Optional[Type]
+
+    def __init__(self, parameters: list[FunctionParameter], result: Optional[Type]):
+        self.parameters = parameters
+        self.result = result
+
+    def is_callable(self, args: Sequence[FunctionArgument]) -> tuple[bool, str | Any]:
+        positionals = [param for param in self.parameters if param.positional]
+        keywords = {param.name: param for param in self.parameters if param.keyword}
+        supplied = {param: False for param in self.parameters}
+        for arg in args:
+            if isinstance(arg.location, int):
+                if arg.location >= len(positionals):
+                    return False, "too many positional arguments supplied"
+                param = positionals[arg.location]
+            else:
+                if arg.location not in keywords:
+                    return False, f"unexpected keyword argument '{arg.location}'"
+                param = keywords[arg.location]
+            if supplied[param]:
+                return False, f"argument '{param.name}' supplied multiple times"
+            if not traits.is_implicitly_convertible(arg.ty, param.ty):
+                return False, f"argument cannot be implicitly converted from '{arg.ty}' to '{param.ty}'  for parameter '{param.name}'"
```

Comment:
In the current type deduction, we provided a full list of which args did or did not match the parameters. I think this is very useful, because this is one of the first errors users run into.

Reply:
We don't currently have a mechanism for displaying batched error messages or notes attached to compiler messages. If we did, I agree that it's a good idea to display error messages eagerly, but until than, I'd rather not make a mess to bolt on a feature that has no proper foundation.